### PR TITLE
Allow sending signals to engines on Keyboard Interrupt in  %%px

### PR DESCRIPTION
**Pull request by sahil1105**

### CODE REVIEW

File: ipyparallel/client/magics.py [hunk 1]
```diff
@@ -364,49 +364,52 @@ def parallel_execute(
...
+                                # already streamed, show an abbreviated result
+                                raise error.AlreadyDisplayedError(e) from None
+                            else:
+                                raise
+                # Skip redisplay if streaming output
+                if not stream_output:
+                    result.display_outputs(groupby)
+            except KeyboardInterrupt:
+                print("Received Keyboard Interrupt. Sending signal 9 to engines...")
+                self.view.client.send_signal(9, targets=targets, block=True)
```
Comment (minrk) [inline ipyparallel/client/magics.py:412]: SIGKILL is a very harsh consequence for impatience. I definitely don't want that to be unconditional. Usually you'd want signal 2 (SIGINT). I realize for MPI that SIGINT isn't always enough, but SIGKILL should be a last resort. Even then, I'm not sure we want it to be unconditional. Maybe a px option, e.g. `--signal-on-interrupt=2|INT|KILL`?

Comment (sahil1105) [reply ipyparallel/client/magics.py:412]: Makes sense. I'll add that option. Should the default be to do nothing (like now), or SIGINT?

Comment (sahil1105) [reply ipyparallel/client/magics.py]: Added the option.

**Comment by minrk** [comment]
I'm not sure about the default. I think sending interrupt by default will _often_ be useful. But in the cases where it's not, it could be pretty destructive. I think for interactivity purposes, let's send SIGINT on interrupt by default (i.e. when we're in a blocking `%px`, SIGINT to the kernel is SIGINT everywhere). I think that matches the trajectory of `%px` creating a 'parallel notebook' where the kernel is a transparent window to the kernels, as much as makes sense.

**Comment by sahil1105** [comment]
> I'm not sure about the default. I think sending interrupt by default will _often_ be useful. But in the cases where it's not, it could be pretty destructive. I think for interactivity purposes, let's send SIGINT on interrupt by default (i.e. when we're in a blocking `%px`, SIGINT to the kernel is SIGINT everywhere). I think that matches the trajectory of `%px` creating a 'parallel notebook' where the kernel is a transparent window to the kernels, as much as makes sense.

Ok, makes sense. Made SIGINT the default, and added a note that 0 acts as a nop.
How should we add a test for this?

**Comment by minrk** [comment]
> How should we add a test for this?

You can try a test with a `signal.alarm` that raises KeyboardInterrupt after a delay, then run a cell that will block for more than a few seconds. It should be skipped on Windows.

**Comment by sahil1105** [comment]
> > How should we add a test for this?
> 
> You can try a test with a `signal.alarm` that raises KeyboardInterrupt after a delay, then run a cell that will block for more than a few seconds. It should be skipped on Windows.

Thanks. Added tests.

**Comment by minrk** [comment]
Thanks!

**Pull request merged**